pkg/prometheus/prometheus.go

```diff
@@ -37,37 +28,102 @@ var ResponseSizeBuckets = append(append(append(append(
 	prometheus.LinearBuckets(1000000, 1000000, 5)...),
 )
 
-// WithTelemetry instruments the HTTP server with prometheus
-func WithTelemetry(handler http.Handler) http.HandlerFunc {
-	counter := prometheus.NewCounterVec(
+var (
+	// server metrics
+	counter = prometheus.NewCounterVec(
 		prometheus.CounterOpts{
 			Name: "http_requests_total",
 			Help: "A counter for requests to the wrapped handler.",
 		},
-		[]string{"code"},
+		[]string{"code", "method"},
 	)
 
-	duration := prometheus.NewHistogramVec(
+	duration = prometheus.NewHistogramVec(
 		prometheus.HistogramOpts{
 			Name:    "http_request_duration_seconds",
 			Help:    "A histogram of latencies for requests in seconds.",
 			Buckets: RequestDurationBucketsSeconds,
 		},
-		[]string{"code"},
+		[]string{"code", "method"},
 	)
 
-	responseSize := prometheus.NewHistogramVec(
+	responseSize = prometheus.NewHistogramVec(
 		prometheus.HistogramOpts{
 			Name:    "http_response_size_bytes",
 			Help:    "A histogram of response sizes for requests.",
 			Buckets: ResponseSizeBuckets,
 		},
-		[]string{},
+		[]string{"code", "method"},
+	)
+
+	// client metrics
+	clientCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "client_api_requests_total",
+			Help: "A counter for requests from the wrapped client.",
+		},
+		[]string{"client", "code", "method"},
+	)
+
+	clientDur = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "client_request_duration_seconds",
+			Help:    "A histogram of request latencies.",
+			Buckets: RequestDurationBucketsSeconds,
+		},
+		[]string{"client", "code", "method"},
+	)
+
+	clientInFlight = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "client_in_flight_requests",
+			Help: "A gauge of in-flight requests for the wrapped client.",
+		},
+		[]string{"client"},
+	)
+)
+
+func init() {
+	prometheus.MustRegister(
+		counter, duration, responseSize, // server metrics
+		clientCounter, clientDur, clientInFlight, // client metrics
 	)
+}
 
-	prometheus.MustRegister(counter, duration, responseSize)
+// NewGrpcServer returns a grpc server pre-configured with prometheus interceptors
+func NewGrpcServer() *grpc.Server {
+	server := grpc.NewServer(
+		grpc.UnaryInterceptor(grpc_prometheus.UnaryServerInterceptor),
+		grpc.StreamInterceptor(grpc_prometheus.StreamServerInterceptor),
+	)
 
+	grpc_prometheus.EnableHandlingTimeHistogram()
+	grpc_prometheus.Register(server)
+	return server
+}
+
+// WithTelemetry instruments the HTTP server with prometheus
+func WithTelemetry(handler http.Handler) http.HandlerFunc {
 	return promhttp.InstrumentHandlerDuration(duration,
 		promhttp.InstrumentHandlerResponseSize(responseSize,
 			promhttp.InstrumentHandlerCounter(counter, handler)))
 }
+
+// ClientWithTelemetry instruments the HTTP client with prometheus
+func ClientWithTelemetry(name string, wt WrapTransport) WrapTransport {
+	dur := clientDur.MustCurryWith(prometheus.Labels{"client": name})
```

Comment:
Why do all these need to panic instead of returning an error?

Reply:
The panics here are intended for development-time validation of our Prometheus labels. A failure at this point in the code indicates that we have misconfigured our Prometheus labels, somewhere between calling `prometheus.NewCounterVec(...` and calling `ClientWithTelemetry`. If our configuration/usage of HTTP clients becomes more dynamic, where unexpected label configurations could arise in production, I'd be inclined to permit an error here.